tao_compiler/mlir/xla/ral/ral_md5.h

```diff
@@ -0,0 +1,116 @@
+Algorithm" in all material mentioning or referencing this software
+or this function.
+
+License is also granted to make and use derivative works provided
+that such works are identified as "derived from the RSA Data
+Security, Inc. MD5 Message-Digest Algorithm" in all material
+mentioning or referencing the derived work.
+
+RSA Data Security, Inc. makes no representations concerning either
+the merchantability of this software or the suitability of this
+software for any particular purpose. It is provided "as is"
+without express or implied warranty of any kind.
+
+These notices must be retained in any copies of any part of this
+documentation and/or software.
+
+*/
+
+#ifndef RAL_RAL_MD5_H_
+#define RAL_RAL_MD5_H_
+
+#include <cstring>
+#include <iostream>
+
+namespace tao {
+namespace ral {
+
+// a small class for calculating MD5 hashes of strings or byte arrays
+// it is not meant to be fast or secure
+//
+// usage: 1) feed it blocks of uchars with update()
+//      2) finalize()
+//      3) get hexdigest() string
+//      or
+//      MD5(std::string).hexdigest()
+//
+// assumes that char is 8 bit and int is 32 bit
+class MD5 {
+ public:
+  typedef unsigned int size_type;  // must be 32bit
+
+  MD5();
+  MD5(const std::string& text);
+  void update(const unsigned char* buf, size_type length);
+  void update(const char* buf, size_type length);
+  MD5& finalize();
+  std::string hexdigest() const;
+  friend std::ostream& operator<<(std::ostream&, MD5 md5);
+
+ private:
+  void init();
+  typedef unsigned char uint1;  //  8bit
+  typedef unsigned int uint4;   // 32bit
+  enum { blocksize = 64 };      // VC6 won't eat a const static int here
+
+  void transform(const uint1 block[blocksize]);
+  static void decode(uint4 output[], const uint1 input[], size_type len);
+  static void encode(uint1 output[], const uint4 input[], size_type len);
+
+  bool finalized;
+  uint1 buffer[blocksize];  // bytes that didn't fit in last 64 byte chunk
+  uint4 count[2];           // 64bit counter for number of bits (lo, hi)
+  uint4 state[4];           // digest so far
+  uint1 digest[16];         // the result
+
+  // low level logic operations
+  static inline uint4 F(uint4 x, uint4 y, uint4 z);
+  static inline uint4 G(uint4 x, uint4 y, uint4 z);
+  static inline uint4 H(uint4 x, uint4 y, uint4 z);
+  static inline uint4 I(uint4 x, uint4 y, uint4 z);
+  static inline uint4 rotate_left(uint4 x, int n);
+  static inline void FF(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
+                        uint4 ac);
+  static inline void GG(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
+                        uint4 ac);
+  static inline void HH(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
+                        uint4 ac);
+  static inline void II(uint4& a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s,
+                        uint4 ac);
+};
+
+std::string md5(const std::string str);
```

Comment:
nit: need a tiny test to guard it?

Reply:
well, it actually copied from a golden implementation.